lib/Sema/ConstraintSystem.h

```diff
@@ -3430,6 +3432,56 @@ class DisjunctionChoice {
   }
 };
 
+class TypeVarBindingGenerator {
```

Comment:
Would it be better to make this look like an STL iterator? begin() end() `operator*` `operator++`

Reply:
I don't really have a preference but @rudkx does and he wanted everything to look more like a generator than iterator, when we discussed last time.

Reply:
I think it makes more sense it this case to keep it closer to generator because we'd want to check if next element need to be computed.

Reply:
I don't have a huge preference either way but I originally imagined making the solver iterative by having a single unified notion of a `ConstraintVariable` that abstracted over type variables and disjunctions and then a generator interface over that abstraction that you could store on a stack which also held the SolverScope that the generator was created in, so that the main solver loop becomes something like:

```
  i = 0
  llvm::SmallVector<SolverStackElement, 16> stack;
  cv = pickNextConstraintVariable()
  cvg = constraintVariableGenerator(cv) // compute the domain, store it and the constraint variable in the generator
  element = SolverStackElement(cvg, SolverScope()) // create a new scope and store it with the variable on 
  stack.push_back(element)

  while (!done) { // where done means running out of constraint variables to try
    cvg = stack[i].generator
    if (bindAndSimplify(cv, cvg)) {
      ++i
      cv = pickNextConstraintVariable()
      cvg = constraintVariableGenerator(cv) // compute the domain, store it and the constraint variable in the generator
      element = SolverStackElement(cvg, SolverScope()) // create a new scope and store it with the variable on 
      stack.push_back(element)
    } else {
      --i
      stack.pop_back()
   }
}
```

where `bindAndSimplify` attempts to bind elements from the domain of the generator until it finds one that `simplify()` doesn't fail on, returning `true` if it finds such a binding, or `false` if they all fail.

Of course this is an oversimplification because it's not considering connected components. It seems like we also need a generator and stack for those.